src/Ql/AbstractQlConnection.php

```diff
@@ -0,0 +1,520 @@
+      },
+      $retries
+    );
+  }
+
+  /**
+   * @param callable      $func
+   * @param callable|null $onError
+   * @param int|null      $retryCount
+   *
+   * @return mixed
+   * @throws ConnectionException
+   */
+  protected function _performWithRetries(
+    callable $func, callable $onError = null, $retryCount = null
+  )
+  {
+    $this->_lastRetryCount = 0;
+    $this->_recycleConnectionIfRequired();
+
+    if($retryCount === null)
+    {
+      $retryCount = (int)$this->_config()->getItem('retries', 2);
+    }
+
+    /** @var null|ConnectionException $exception */
+    $exception = null;
+    $retries = $retryCount;
+    do
+    {
+      try
+      {
+        $this->_lastRetryCount++;
+        return $func();
+      }
+      catch(\Exception $sourceException)
+      {
+        if($onError)
+        {
+          $onError();
+        }
+
+        $exception = $this->_sanitizeException($sourceException);
+        $recoverable = $this->_isRecoverableException($exception);
+        if($retries > 0 && $recoverable)
+        {
+          if($this->_shouldReconnectAfterException($exception))
+          {
+            if($this->_inTransaction)
+            {
+              error_log(
+                'Connection error during transaction: '
+                . '(' . $exception->getCode() . ') ' . $exception->getMessage()
+              );
+              throw $exception;
+            }
+            $this->disconnect();
+          }
+
+          //Sleep 1ms > 30ms
+          usleep(mt_rand(1000, 30000));
+          $this->connect();
+        }
+        else
+        {
+          $this->disconnect();
+          error_log(
+            'Connection Error: (' . $exception->getCode() . ') '
+            . $exception->getMessage()
+          );
+          throw $exception;
+        }
+      }
+      $retries--;
+    }
+    while($retries > 0);
+
+    if($exception)
+    {
+      throw $exception;
+    }
+    else
+    {
+      throw new ConnectionException(
+        'An unknown error occurred performing an operation. '
+        . 'The operation failed after ' . $retryCount . ' retries'
+      );
+    }
+  }
+
+  /**
+   * @param \Exception $e
+   *
+   * @return \Exception
+   */
+  protected function _sanitizeException(\Exception $e)
+  {
+    return ConnectionException::from($e);
+  }
+
+  /**
+   * @param \Exception $e
+   *
+   * @return bool
+   */
+  abstract protected function _isRecoverableException(\Exception $e);
+
+  /**
+   * Should we reconnect to the database after this sort of error?
+   *
+   * @param \Exception $e
+   *
+   * @return bool
+   */
+  private function _shouldReconnectAfterException(\Exception $e)
+  {
+    // 2006  = MySQL server has gone away
+    // 1047  = ER_UNKNOWN_COM_ERROR - happens when a PXC node is resyncing:
+    //          "WSREP has not yet prepared node for application use"
+    // HY000 = General SQL error
+    $codes = ['2006', '1047', '2014', 'HY000'];
```

Comment:
What is code 2014? Can you add it to the comment so we don't forget?

Reply:
good catch, i think that one is a mistake. It's "bad query", we don't need to reconnect for that.